File: freehold/core/control_flow.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from freehold.core.ast import *


@dataclass(frozen=True)
class RoutineFlowSummary:
    routine_name: str
    routine_kind: str
    normal_return_possible: bool
    guaranteed_exit: bool
    declared_aborts: frozenset[str]
    emitted_aborts: frozenset[str]
    called_routines: frozenset[str]
    propagated_aborts: frozenset[str]


@dataclass(frozen=True)
class _BlockFlow:
    normal_return_possible: bool
    guaranteed_exit: bool
    emitted_aborts: frozenset[str]
    called_routines: frozenset[str]


class ControlFlowAnalyzer:
    def __init__(self, routines: dict[str, RoutineDecl], module_name: str = ""):
        self.routines = routines
        self.module_name = module_name

    def analyze_program(self, program: Program) -> dict[str, RoutineFlowSummary]:
        analyzer = ControlFlowAnalyzer(
            {declaration.name: declaration for declaration in program.declarations if isinstance(declaration, RoutineDecl)},
            program.module_name,
        )
        return analyzer.analyze_routines()

    def analyze_routines(self) -> dict[str, RoutineFlowSummary]:
        return {name: self.analyze_routine(routine) for name, routine in self.routines.items()}

    def analyze_routine(self, routine: RoutineDecl) -> RoutineFlowSummary:
        body_flow = self._block(routine.body)
        declared_aborts = frozenset(clause.error_name for clause in routine.aborts)
        normal_return_possible = body_flow.normal_return_possible or (
            routine.kind == "procedure" and not body_flow.guaranteed_exit
        )
        return RoutineFlowSummary(
            routine_name=routine.name,
            routine_kind=routine.kind,
            normal_return_possible=normal_return_possible,
            guaranteed_exit=body_flow.guaranteed_exit,
            declared_aborts=declared_aborts,
            emitted_aborts=body_flow.emitted_aborts,
            called_routines=body_flow.called_routines,
            propagated_aborts=declared_aborts,
        )

    def _block(self, body: list[Any]) -> _BlockFlow:
        normal_return_possible = False
        emitted_aborts: set[str] = set()
        called_routines: set[str] = set()

        for statement in body:
            statement_flow = self._statement(statement)
            normal_return_possible = normal_return_possible or statement_flow.normal_return_possible
            emitted_aborts.update(statement_flow.emitted_aborts)
            called_routines.update(statement_flow.called_routines)
            if statement_flow.guaranteed_exit:
                return _BlockFlow(
                    normal_return_possible,
                    True,
                    frozenset(emitted_aborts),
                    frozenset(called_routines),
                )

        return _BlockFlow(normal_return_possible, False, frozenset(emitted_aborts), frozenset(called_routines))

    def _statement(self, statement: Any) -> _BlockFlow:
        if isinstance(statement, ReturnStmt):
            called_routines = self._calls_in_expr(statement.value)
            return _BlockFlow(True, True, frozenset(), frozenset(called_routines))
        if isinstance(statement, AbortStmt):
            return _BlockFlow(False, True, frozenset({statement.error_name}), frozenset())
        if isinstance(statement, CallStmt):
            called_routines = set(self._routine_call_name(statement.name))
            for argument in statement.args:
                called_routines.update(self._calls_in_expr(argument))
            return _BlockFlow(False, False, frozenset(), frozenset(called_routines))
        if isinstance(statement, LetStmt):
            return self._expression_statement_flow(statement.expr)
        if isinstance(statement, AssignStmt):
            return self._expression_statement_flow(statement.expr)
        if isinstance(statement, FieldAssignStmt):
            return self._expression_statement_flow(statement.expr)
        if isinstance(statement, CheckStmt):
            return self._expression_statement_flow(statement.expr)
        if isinstance(statement, IfStmt):
            condition_calls = self._calls_in_expr(statement.condition)
            then_flow = self._block(statement.then_body)
            else_flow = self._block(statement.else_body)
            return _BlockFlow(
                then_flow.normal_return_possible or else_flow.normal_return_possible,
                then_flow.guaranteed_exit and else_flow.guaranteed_exit,
                then_flow.emitted_aborts | else_flow.emitted_aborts,
                frozenset(set(condition_calls) | set(then_flow.called_routines) | set(else_flow.called_routines)),
            )
        if isinstance(statement, WhileStmt):
            body_flow = self._block(statement.body)
            called_routines = set(self._calls_in_expr(statement.condition)) | set(body_flow.called_routines)
            for invariant in statement.invariants:
                called_routines.update(self._calls_in_expr(invariant))
            if statement.variant is not None:
                called_routines.update(self._calls_in_expr(statement.variant))
            return _BlockFlow(False, False, body_flow.emitted_aborts, frozenset(called_routines))
        if isinstance(statement, CaseStmt):
            case_calls = set(self._calls_in_expr(statement.expr))
            branch_flows = [self._block(branch.body) for branch in statement.branches]
            default_flow = self._block(statement.default_body)
            all_flows = branch_flows + [default_flow]
            emitted_aborts: set[str] = set()
            called_routines: set[str] = set(case_calls)
            normal_return_possible = False
            guaranteed_exit = True
            for branch in statement.branches:
                called_routines.update(self._calls_in_expr(branch.value))
            for flow in all_flows:
                emitted_aborts.update(flow.emitted_aborts)
                called_routines.update(flow.called_routines)
                normal_return_possible = normal_return_possible or flow.normal_return_possible
                guaranteed_exit = guaranteed_exit and flow.guaranteed_exit
            return _BlockFlow(normal_return_possible, guaranteed_exit, frozenset(emitted_aborts), frozenset(called_routines))
        return _BlockFlow(False, False, frozenset(), frozenset())

    def _expression_statement_flow(self, expression: Any) -> _BlockFlow:
        return _BlockFlow(False, False, frozenset(), frozenset(self._calls_in_expr(expression)))

    def _calls_in_expr(self, expression: Any) -> set[str]:
        if expression is None:
            return set()
        if isinstance(expression, (ReturnPlain, ReturnOk)):
            return self._calls_in_expr(expression.expr)
        if isinstance(expression, ReturnError):
            return set()
        if isinstance(expression, CallExpr):
            called_routines = set(self._routine_call_name(expression.name))
            for argument in expression.args:
                called_routines.update(self._calls_in_expr(argument))
            return called_routines
        if isinstance(expression, NamedArg):
            return self._calls_in_expr(expression.expr)
        if isinstance(expression, RecordLiteralExpr):
            called_routines: set[str] = set()
            for argument in expression.args:
                called_routines.update(self._calls_in_expr(argument))
            return called_routines
        if isinstance(expression, ArrayLiteralExpr):
            called_routines: set[str] = set()
            for item in expression.items:
                called_routines.update(self._calls_in_expr(item))
            return called_routines
        if isinstance(expression, IndexExpr):
            return self._calls_in_expr(expression.index)
        if isinstance(expression, UnaryExpr):
            return self._calls_in_expr(expression.expr)
        if isinstance(expression, BinaryExpr):
            return self._calls_in_expr(expression.left) | self._calls_in_expr(expression.right)
        return set()

    def _routine_call_name(self, name: str) -> set[str]:
        local_name = self._local_routine_name(name)
        if local_name in self.routines:
            return {local_name}
        return set()

    def _local_routine_name(self, name: str) -> str:
        prefix = f"{self.module_name}."
        return name[len(prefix):] if self.module_name and name.startswith(prefix) else name

File: go-frontend/internal/semantic/control_flow.go
package semantic

import (
	"strings"

	"freehold-go-frontend/internal/ast"
)

type blockFlow struct {
	normalReturnPossible bool
	guaranteedExit       bool
	emittedAborts        map[string]bool
	calledRoutines       map[string]bool
}

type ControlFlowAnalyzer struct {
	routines   map[string]ast.Decl
	moduleName string
}

func NewControlFlowAnalyzer(module *ast.Module) *ControlFlowAnalyzer {
	routines := make(map[string]ast.Decl)
	if module != nil {
		for _, decl := range module.Declarations {
			switch d := decl.(type) {
			case ast.FunctionDecl:
				routines[d.Name] = d
			case ast.ProcedureDecl:
				routines[d.Name] = d
			}
		}
	}
	var moduleName string
	if module != nil {
		moduleName = module.Name
	}
	return &ControlFlowAnalyzer{
		routines:   routines,
		moduleName: moduleName,
	}
}

func (c *ControlFlowAnalyzer) AnalyzeRoutines() map[string]ast.RoutineFlowSummary {
	summaries := make(map[string]ast.RoutineFlowSummary)
	for name, decl := range c.routines {
		summaries[name] = c.AnalyzeRoutine(decl)
	}
	return summaries
}

func (c *ControlFlowAnalyzer) AnalyzeRoutine(decl ast.Decl) ast.RoutineFlowSummary {
	var body []ast.Stmt
	var kind string
	var name string
	var aborts []ast.AbortClause

	switch d := decl.(type) {
	case ast.FunctionDecl:
		body = d.Body
		kind = "function"
		name = d.Name
		aborts = d.Aborts
	case ast.ProcedureDecl:
		body = d.Body
		kind = "procedure"
		name = d.Name
		aborts = d.Aborts
	}

	bodyFlow := c.analyzeBlock(body)
	declaredAborts := make(map[string]bool)
	for _, clause := range aborts {
		declaredAborts[clause.Error] = true
	}

	normalReturnPossible := bodyFlow.normalReturnPossible || (kind == "procedure" && !bodyFlow.guaranteedExit)

	// In V0, propagated aborts is equivalent to declared aborts
	propagatedAborts := make(map[string]bool)
	for k, v := range declaredAborts {
		propagatedAborts[k] = v
	}

	return ast.RoutineFlowSummary{
		RoutineName:          name,
		RoutineKind:          kind,
		NormalReturnPossible: normalReturnPossible,
		GuaranteedExit:       bodyFlow.guaranteedExit,
		DeclaredAborts:       declaredAborts,
		EmittedAborts:        bodyFlow.emittedAborts,
		CalledRoutines:       bodyFlow.calledRoutines,
		PropagatedAborts:     propagatedAborts,
	}
}

func (c *ControlFlowAnalyzer) analyzeBlock(body []ast.Stmt) blockFlow {
	normalReturnPossible := false
	emittedAborts := make(map[string]bool)
	calledRoutines := make(map[string]bool)

	for _, stmt := range body {
		stmtFlow := c.analyzeStatement(stmt)
		if stmtFlow.normalReturnPossible {
			normalReturnPossible = true
		}
		for k, v := range stmtFlow.emittedAborts {
			emittedAborts[k] = v
		}
		for k, v := range stmtFlow.calledRoutines {
			calledRoutines[k] = v
		}
		if stmtFlow.guaranteedExit {
			return blockFlow{
				normalReturnPossible: normalReturnPossible,
				guaranteedExit:       true,
				emittedAborts:        emittedAborts,
				calledRoutines:       calledRoutines,
			}
		}
	}

	return blockFlow{
		normalReturnPossible: normalReturnPossible,
		guaranteedExit:       false,
		emittedAborts:        emittedAborts,
		calledRoutines:       calledRoutines,
	}
}

func (c *ControlFlowAnalyzer) analyzeStatement(stmt ast.Stmt) blockFlow {
	switch s := stmt.(type) {
	case ast.ReturnStmt:
		calls := c.callsInExpr(s.Value)
		return blockFlow{
			normalReturnPossible: true,
			guaranteedExit:       true,
			emittedAborts:        make(map[string]bool),
			calledRoutines:       calls,
		}
	case ast.AbortStmt:
		emitted := map[string]bool{s.Error: true}
		return blockFlow{
			normalReturnPossible: false,
			guaranteedExit:       true,
			emittedAborts:        emitted,
			calledRoutines:       make(map[string]bool),
		}
	case ast.CallStmt:
		calls := make(map[string]bool)
		if callExpr, ok := s.Call.(ast.CallExpr); ok {
			for k, v := range c.routineCallName(callExpr.Callee) {
				calls[k] = v
			}
			for _, arg := range callExpr.Arguments {
				for k, v := range c.callsInExpr(arg) {
					calls[k] = v
				}
			}
		} else {
			for k, v := range c.callsInExpr(s.Call) {
				calls[k] = v
			}
		}
		return blockFlow{
			normalReturnPossible: false,
			guaranteedExit:       false,
			emittedAborts:        make(map[string]bool),
			calledRoutines:       calls,
		}
	case ast.LetStmt:
		return blockFlow{
			normalReturnPossible: false,
			guaranteedExit:       false,
			emittedAborts:        make(map[string]bool),
			calledRoutines:       c.callsInExpr(s.Value),
		}
	case ast.AssignmentStmt:
		calls := c.callsInExpr(s.Value)
		for k, v := range c.callsInExpr(s.Target) {
			calls[k] = v
		}
		return blockFlow{
			normalReturnPossible: false,
			guaranteedExit:       false,
			emittedAborts:        make(map[string]bool),
			calledRoutines:       calls,
		}
	case ast.CheckStmt:
		return blockFlow{
			normalReturnPossible: false,
			guaranteedExit:       false,
			emittedAborts:        make(map[string]bool),
			calledRoutines:       c.callsInExpr(s.Condition),
		}
	case ast.IfStmt:
		conditionCalls := c.callsInExpr(s.Condition)
		thenFlow := c.analyzeBlock(s.ThenBody)
		elseFlow := c.analyzeBlock(s.ElseBody)

		emitted := make(map[string]bool)
		for k, v := range thenFlow.emittedAborts {
			emitted[k] = v
		}
		for k, v := range elseFlow.emittedAborts {
			emitted[k] = v
		}

		calls := make(map[string]bool)
		for k, v := range conditionCalls {
			calls[k] = v
		}
		for k, v := range thenFlow.calledRoutines {
			calls[k] = v
		}
		for k, v := range elseFlow.calledRoutines {
			calls[k] = v
		}

		return blockFlow{
			normalReturnPossible: thenFlow.normalReturnPossible || elseFlow.normalReturnPossible,
			guaranteedExit:       thenFlow.guaranteedExit && elseFlow.guaranteedExit,
			emittedAborts:        emitted,
			calledRoutines:       calls,
		}
	case ast.WhileStmt:
		bodyFlow := c.analyzeBlock(s.Body)
		calls := c.callsInExpr(s.Condition)
		for k, v := range bodyFlow.calledRoutines {
			calls[k] = v
		}
		for _, invariant := range s.Invariants {
			for k, v := range c.callsInExpr(invariant) {
				calls[k] = v
			}
		}
		if s.Variant != nil {
			for k, v := range c.callsInExpr(s.Variant) {
				calls[k] = v
			}
		}
		return blockFlow{
			normalReturnPossible: false,
			guaranteedExit:       false,
			emittedAborts:        bodyFlow.emittedAborts,
			calledRoutines:       calls,
		}
	case ast.CaseStmt:
		caseCalls := c.callsInExpr(s.Value)
		var branchFlows []blockFlow
		for _, branch := range s.When {
			branchFlows = append(branchFlows, c.analyzeBlock(branch.Body))
		}
		defaultFlow := c.analyzeBlock(s.Default)

		emitted := make(map[string]bool)
		calls := make(map[string]bool)
		for k, v := range caseCalls {
			calls[k] = v
		}

		for _, branch := range s.When {
			for k, v := range c.callsInExpr(branch.Value) {
				calls[k] = v
			}
		}

		normalReturnPossible := false
		guaranteedExit := true

		allFlows := append(branchFlows, defaultFlow)
		for _, flow := range allFlows {
			for k, v := range flow.emittedAborts {
				emitted[k] = v
			}
			for k, v := range flow.calledRoutines {
				calls[k] = v
			}
			if flow.normalReturnPossible {
				normalReturnPossible = true
			}
			if !flow.guaranteedExit {
				guaranteedExit = false
			}
		}

		return blockFlow{
			normalReturnPossible: normalReturnPossible,
			guaranteedExit:       guaranteedExit,
			emittedAborts:        emitted,
			calledRoutines:       calls,
		}
	case ast.ScopeStmt:
		// A scope runs SpawnBody, JoinBody, then ResultBody in sequence
		spawnFlow := c.analyzeBlock(s.SpawnBody)
		joinFlow := c.analyzeBlock(s.JoinBody)
		resultFlow := c.analyzeBlock(s.ResultBody)

		emitted := make(map[string]bool)
		calls := make(map[string]bool)

		// Accumulate emitted aborts
		for k, v := range spawnFlow.emittedAborts {
			emitted[k] = v
		}
		for k, v := range joinFlow.emittedAborts {
			emitted[k] = v
		}
		for k, v := range resultFlow.emittedAborts {
			emitted[k] = v
		}

		// Accumulate called routines
		for k, v := range spawnFlow.calledRoutines {
			calls[k] = v
		}
		for k, v := range joinFlow.calledRoutines {
			calls[k] = v
		}
		for k, v := range resultFlow.calledRoutines {
			calls[k] = v
		}

		normalReturnPossible := spawnFlow.normalReturnPossible || joinFlow.normalReturnPossible || resultFlow.normalReturnPossible
		guaranteedExit := spawnFlow.guaranteedExit || joinFlow.guaranteedExit || resultFlow.guaranteedExit

		return blockFlow{
			normalReturnPossible: normalReturnPossible,
			guaranteedExit:       guaranteedExit,
			emittedAborts:        emitted,
			calledRoutines:       calls,
		}
	}

	return blockFlow{
		normalReturnPossible: false,
		guaranteedExit:       false,
		emittedAborts:        make(map[string]bool),
		calledRoutines:       make(map[string]bool),
	}
}

func (c *ControlFlowAnalyzer) callsInExpr(expr ast.Expr) map[string]bool {
	calls := make(map[string]bool)
	if expr == nil {
		return calls
	}

	switch e := expr.(type) {
	case ast.OkExpr:
		return c.callsInExpr(e.Value)
	case ast.CallExpr:
		for k, v := range c.routineCallName(e.Callee) {
			calls[k] = v
		}
		for _, arg := range e.Arguments {
			for k, v := range c.callsInExpr(arg) {
				calls[k] = v
			}
		}
	case ast.NamedArgumentExpr:
		return c.callsInExpr(e.Value)
	case ast.RecordLiteralExpr:
		for _, field := range e.Fields {
			for k, v := range c.callsInExpr(field.Value) {
				calls[k] = v
			}
		}
	case ast.ArrayLiteralExpr:
		for _, elem := range e.Elements {
			for k, v := range c.callsInExpr(elem) {
				calls[k] = v
			}
		}
	case ast.IndexExpr:
		for k, v := range c.callsInExpr(e.Array) {
			calls[k] = v
		}
		for k, v := range c.callsInExpr(e.Index) {
			calls[k] = v
		}
	case ast.UnaryExpr:
		return c.callsInExpr(e.Value)
	case ast.BinaryExpr:
		for k, v := range c.callsInExpr(e.Left) {
			calls[k] = v
		}
		for k, v := range c.callsInExpr(e.Right) {
			calls[k] = v
		}
	case ast.AwaitExpr:
		return c.callsInExpr(e.Value)
	}

	return calls
}

func (c *ControlFlowAnalyzer) routineCallName(expr ast.Expr) map[string]bool {
	calls := make(map[string]bool)
	name, ok := callName(expr)
	if !ok {
		return calls
	}

	localName := c.localRoutineName(name)
	if _, exists := c.routines[localName]; exists {
		calls[localName] = true
	}
	return calls
}

func (c *ControlFlowAnalyzer) localRoutineName(name string) string {
	if c.moduleName == "" {
		return name
	}
	prefix := c.moduleName + "."
	if strings.HasPrefix(name, prefix) {
		return name[len(prefix):]
	}
	return name
}
